### PR TITLE
Adding a 'Typename' strongtype for representing all resource types

### PR DESCRIPTION
**Pull request by vishh**
For #168 and #1519.

### CODE REVIEW

**Comment by erictune** [comment]
Need to disambiguate between compute resource and REST resource in the comments.

**Comment by thockin** [comment]
Is pkg/types better, or is this tightly bound to resources?

**Comment by vishh** [comment]
@erictune: Done. PTAL.
@thockin: Do you foresee using this naming pattern for any other purpose?

File: pkg/api/resource/typename_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"testing"
+)
+
+func TestStandardResourceTypes(t *testing.T) {
+	table := []struct {
+		input  string
+		result string
+	}{
+		{"memory", "kubernetes.io/resources/memory"},
```
Comment (bgrant0607) [inline pkg/api/resource/typename_test.go]: Sorry. Somewhere we discussed making this consistent with labels and annotations: `resources.kubernetes.io/memory`.

Comment (bgrant0607) [inline pkg/api/resource/typename_test.go]: Ah: #2565.

/cc @thockin 

Also see:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/identifiers.md

Comment (bgrant0607) [inline pkg/api/resource/typename_test.go]: It would be useful to update resources.md, also.

File: pkg/api/resource/typename.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+//
+type Typename struct {
+	// Fully qualified compute resource type name.
+	Name string
+}
+
+const (
+	// The default compute resource namespace for all standard resource types in kubernetes.
+	DefaultNamespace  = "kubernetes.io/resources"
+	pathCharsREstring = "[a-zA-Z0-9\\-_~!$&\\'()*+,;=:@]+"
```
Comment (bgrant0607) [inline pkg/api/resource/typename.go]: We should use validation functions from:
https://github.com/GoogleCloudPlatform/kubernetes/blob/7f2d0c0f710617ef1f5eec4745b23e0d3f360037/pkg/util/validation.go#L68

To ensure consistency. (Create new ones there if necessary.)

File: pkg/api/resource/typename.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+// "kubernetes.io/resources/" in order to fully-qualify it.
+// This namespace is reserved for code in the open source Kubernetes repository.
+// As a result, all user typenames *MUST* be fully qualified, and cannot be created in this namespace.
+//
+// The recommended best practice is to use a lowercase subdomain and Go-like ASCII camelCase
+// for path components. Some example typenames include "memory"
+// (which will be fully-qualified as "kubernetes.io/resources/memory"),
+// and "myveryown.org/shinyNewResource/v2".
+//
+type Typename struct {
```
Comment (bgrant0607) [inline pkg/api/resource/typename.go]: If this is going to be exposed via the API, it needs to go in api/types.go and api/v1beta3/types.go.

**Comment by bgrant0607** [comment]
pkg/types seems reasonable to me. The same code should apply to labels and annotations. 

Label validation is currently in:
pkg/api/validation/validation.go

which calls parsing code in:
pkg/util/validation.go

Can this be merged with that?

File: pkg/api/resource/typename.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// It helps enforce a specific format for naming compute resource types by providing
+// convinient input parsing methods.
+//
+// The naming format for a compute resource type is described below:
+//
+// A fully-qualified compute resource typename is constructed from a
+// DNS-style subdomain with at least one dot, a slash /, and
+// a path comprised of one or more segments separated by slashes.
+// Specifically,
+// 1. The subdomain must conform to RFC 1035 section 2.3.1 'subdomain' syntax
```
Comment (thockin) [inline pkg/api/resource/typename.go]: Yeah, we should just reduce this all to the format of "qualified name" - "kubernetes.io/memory".  I can pull together a cleanup PR or you can do it - your choice.

File: pkg/api/resource/typename.go (hunk 3 above)
Comment (thockin) [inline pkg/api/resource/typename.go]: Why a struct rather than an alias to string?

File: pkg/api/resource/typename.go (hunk 4 above)
Comment (vishh) [inline pkg/api/resource/typename.go]: Brian clarified the format offline. I will make the changes in this PR.

File: pkg/api/resource/typename_test.go (hunk 1 above)
Comment (vishh) [inline pkg/api/resource/typename_test.go]: Sent #3748.

File: pkg/api/resource/typename.go (hunk 2 above)
Comment (vishh) [inline pkg/api/resource/typename.go]: Done.

File: pkg/api/resource/typename.go (hunk 3 above)
Comment (vishh) [inline pkg/api/resource/typename.go]: It can be an alias. Done.

**Comment by vishh** [comment]
@bgrant0607, @thockin: Addressed all comments. PTAL

File: pkg/api/helpers.go [hunk 5]
```diff
@@ -84,3 +85,21 @@ func IsPullIfNotPresent(p PullPolicy) bool {
...
+	switch str {
+	case string(ResourceMemory), string(ResourceCPU):
+		return true
+	}
+
+	return false
+}
+
+// Returns fully qualified name of a compute resource.
+func GetQualifiedResourceName(resourceName string) ResourceName {
```
Comment (thockin) [inline pkg/api/helpers.go]: Can you move this to util and call it SplitQualifiedName() (return namespace and name), then set the default, then rejoin with JoinQualifiedName() ?

File: pkg/api/types.go [hunk 6]
```diff
@@ -752,6 +752,8 @@ type NodeResources struct {
 type ResourceName string
 
 const (
+	// The default compute resource namespace for all standard resource types.
+	DefaultResourceNamespace = "resources.kubernetes.io"
```
Comment (thockin) [inline pkg/api/types.go]: propose just kubernetes.io ?

File: pkg/api/helpers.go [hunk 7]
```diff
@@ -84,3 +85,21 @@ func IsPullIfNotPresent(p PullPolicy) bool {
 func pullPoliciesEqual(p1, p2 PullPolicy) bool {
 	return strings.ToLower(string(p1)) == strings.ToLower(string(p2))
 }
+
+func IsStandardResourceName(str string) bool {
```
Comment (thockin) [inline pkg/api/helpers.go]: why not init a StringSet with the valid value, then just test .Has() ?

File: pkg/api/helpers.go (hunk 5 above)
Comment (vishh) [inline pkg/api/helpers.go]: Done. I will implement the default namespacing part in another PR.

File: pkg/api/types.go (hunk 6 above)
Comment (vishh) [inline pkg/api/types.go]: Waiting for a conclusion on this on #3748

File: pkg/api/helpers.go (hunk 7 above)
Comment (vishh) [inline pkg/api/helpers.go]: Din't know about StringSet. Done.

File: pkg/api/validation/validation_test.go [hunk 8]
```diff
@@ -1440,3 +1440,40 @@ func TestValidateMinionUpdate(t *testing.T) {
...
+		success bool
+	}{
+		{"memory", true},
+		{"cpu", true},
+		{"network", false},
+		{"disk", false},
+		{"", false},
+		{".", false},
+		{"..", false},
+		{"resources.kubernetes.io/cpu", true},
```
Comment (bgrant0607) [inline pkg/api/validation/validation_test.go]: Remove resources., as per resources.md.

Comment (vishh) [inline pkg/api/validation/validation_test.go]: Done

**Comment by thockin** [comment]
LGTM

**Pull request merged**